### DISCUSSION TITLE
IWorkflowInvoker not found in Elsa.Services

### DISCUSSION THREAD

**Discussion by makattar** [elsa-workflows/elsa-core#1198] [Q&A]
Hi .
I am using Elsa 2.0.0.159 .I have embedded elsa in .Net core 5 Web App .I have designed a workflow using Signal Activity in designer .Now when i get a request to my .net controller action there i wants to invoke that workflow earlier designed in designer.I tried to implement using IWorkflowInvoker but IWorkflowInvoker isn't found.Please help me to get out .
Thank you.

using Elsa.WebApp.ProjectOne.Models;
using Elsa.WebApp.ProjectOne.Services.Interfaces;
using Microsoft.Extensions.Configuration;
using MongoDB.Driver;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Elsa.WebApp.ProjectOne.DataLayer;
using Elsa.Activities.Workflows;
using MongoDB.Bson;
using Elsa.Services;
using Microsoft.Extensions.DependencyInjection;
using Elsa.WebApp.ProjectOne.Workflows;
using Elsa.Services.Models;
using Elsa.Activities.Workflows.Extensions;
using Elsa.Models;

namespace Elsa.WebApp.ProjectOne.Services.Classes
{
    public class EmployeeService : IEmployeeService
    {
        private readonly MongoRepository _repository = null;
        private  IServiceProvider _serviceProvider;
        public EmployeeService(IConfiguration configuration,IServiceProvider serviceProvider)
        {
            _repository = new MongoRepository(configuration);
            _serviceProvider = serviceProvider;
        }
        
        public async Task<RunWorkflowResult> TriggerWorkFlow()
        {
            //Wants to Execute Workflow Designed in designer here
           
        }
    }
}

**Comment by sfmskywalker** [comment]
`IWorkflowInvoker` has been replaced with `IWorkflowRunner` and a number of high-level services such as `IWorkflowLaunchpad` to deal with querying what workflows to trigger.

To run a given workflow, you need to:

1. Get the workflow blueprint from `IWorkflowRegistry`.
2. Create a workflow instance.
3. Invoke `IWorkflowRunner.RunWorkflowAsync`.

The `IWorkflowLaunchpad` makes wraps these steps using a single call:

```csharp
_workflowLaunchpad.CollectAndExecuteStartableWorkflowAsync(workflowDefinitionId);
```

**Reply by fulganta** [reply]
I used on Elsa 1.5.3 the following that is not working anymore with Elsa 2.2.1.
I was triggering an aync worflow starting by a signal activity that was waiting for "OnPriceRequestReceived"
```
using Elsa.Activities.Workflows.Extensions;
using Elsa.Models;
using Elsa.Services;

public async void TriggerDefaultPricingWorkFlow([FromBody] QuoteRequest quoteRequest)
{
   try
	{
         var input = new Variables();
         input.SetVariable("QuoteRequest", quoteRequest);
         await workFlowInvoker.TriggerSignalAsync("OnPriceRequestReceived", input, correlationId: quoteRequest.CorrelationId);
```

Do you have an example of how to do it with `IWorkflowRunner `and `IWorkflowRegistry `or the `IWorkflowLaunchpad `you mention above?

**Reply by sfmskywalker** [reply]
Sure. Since you're talking specifically about triggering a signal activity, the easiest way is to use `ISignaler` like this:

```csharp
await _signaler.TriggerSignalAsync("OnPriceRequestReceived", input: quoteRequest, correlationId: quoteRequest.CorrelationId);
```

Under the hood, it uses `IWorkflowLaunchpad` like this:

```csharp
public async Task<IEnumerable<CollectedWorkflow>> TriggerSignalAsync(string signal, object? input = default, string? workflowInstanceId = default, string? correlationId = default, CancellationToken cancellationToken = default)
{
    var normalizedSignal = signal.ToLowerInvariant();

    return await _workflowLaunchpad.CollectAndExecuteWorkflowsAsync(new WorkflowsQuery(
        nameof(SignalReceived),
        new SignalReceivedBookmark { Signal = normalizedSignal },
        correlationId,
        workflowInstanceId,
        default,
        TenantId
    ), new WorkflowInput(new Signal(normalizedSignal, input)), cancellationToken);
}
```

**Reply by fulganta** [reply]
Hi,

Thanks again for your quick feedback. sorry for my stupid questions as I'm not super familiar yet with C# itself...

Basically what I'm trying to achieve is to have two app:
- 1 swagger Controller ASP.NET that receive diverse requests.
- 1 Elsa orchestrator app that will handle pure workflow requests.

The swagger will occasionally trigger an Elsa workflow and call Elsa app.
 My plan was either to trigger the workflow from the swagger app
1- By sending a post + input data to start the workflow calling an Elsa http Endpoint activity.
2- or I was also exploring having the swagger app using Elsa libraries itself to trigger the elsa workflow using the  _signaler.TriggerSignalAsync function. 

The first one I succeed to make it. The second one I have difficulty as exposed above. If you have any comment on what make more sense let me know betwen 1 and 2. I think first one make more sense now as I realize that it requires deploying too many Elsa lib and starting another Elsa server mode on the swagger app which is not super smart.

But anyhow if I do that and in my swagger Startup.cs if include only my dependency injection 
`services.AddTransient<ISignaler, Signaler>();` in order to call the ISignaler object in my controller constructor
I will get some errors:

```
System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Elsa.Activities.Signaling.Services.ISignaler Lifetime: Transient ImplementationType: Elsa.Activities.Signaling.Services.Signaler': Unable to resolve service for type 'Elsa.Services.IWorkflowLaunchpad' while attempting to activate 'Elsa.Activities.Signaling.Services.Signaler'.)'
InvalidOperationException: Unable to resolve service for type 'Elsa.Services.IWorkflowLaunchpad' while attempting to activate 'Elsa.Activities.Signaling.Services.Signaler'.
```

So I believe I can't only declare the dependency injection, i need to start also an Elsa server in my startup.cs from my swagger project.

But if I add

      public void ConfigureServices(IServiceCollection services)
      {
      .
      .
      .
      var elsaSection = Configuration.GetSection("Elsa");
      
      //Elsa services.
      services
      .AddElsa(elsa => elsa
	      .UseEntityFrameworkPersistence(ef => ef.UseSqlite())
	      .AddConsoleActivities()
	      .AddHttpActivities(elsaSection.GetSection("Server").Bind)
	      .AddQuartzTemporalActivities()
	      .AddWorkflowsFrom<Startup>()
      );
      
      //Elsa API endpoints.
	      services.AddElsaApiEndpoints();
      
      //Register all Mediatr event handlers from this assembly.
	      services.AddNotificationHandlersFrom<Startup>();

it will this time break if I add `endpoints.MapControllers();`:
```
app.UseEndpoints(endpoints =>
            {
				//endpoints.MapDefaultControllerRoute();
				endpoints.MapControllers();
				endpoints.MapHub<SignalrNotificationHub>("/resultnotificationhub");
			});
```

With error:
```
System.Reflection.ReflectionTypeLoadException: 'Unable to load one or more of the requested types.
Could not load type 'Elsa.Metadata.ActivityDescriptor' from assembly 'Elsa.Core, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.
Could not load type 'Elsa.Services.Models.IActivity' from assembly 'Elsa.Abstractions, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.
Could not load type 'Elsa.Models.WorkflowDefinitionVersion' from assembly 'Elsa.Abstractions, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.
Could not load type 'Elsa.Models.WorkflowDefinitionVersion' from assembly 'Elsa.Abstractions, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.
Could not load type 'Elsa.Models.WorkflowDefinitionVersion' from assembly 'Elsa.Abstractions, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.
Could not load type 'Elsa.Models.WorkflowDefinitionVersion' from assembly 'Elsa.Abstractions, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.
Could not load type 'Elsa.Models.WorkflowDefinitionVersion' from assembly 'Elsa.Abstractions, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.
Could not load type 'Elsa.Models.WorkflowDefinitionVersion' from assembly 'Elsa.Abstractions, Version=2.2.1.0, Culture=neutral, PublicKeyToken=null'.'
```

If I don't add 
`endpoints.MapControllers();

It just broke my swagger 
![image](https://user-images.githubusercontent.com/45927310/134287169-5bccdcd1-a739-4496-a263-9f749203cb02.png)

So I believe I'm probably trying to register tow services on same port or something like that and I broke it.

I'm not sure what's the best approach here. sorry to bother you with that,

**Reply by sfmskywalker** [reply]
Regarding strategy 1 vs 2, either way is fine. It all depends on your requirements.

For some apps, it makes sense to execute workflows within the same application. A good example is Orchard Core, an application framework with on top of that CMS functionality. The idea is that workflows allow users to respond to events happening within the system.

For other apps, it makes sense to execute workflows external to other apps. For example, microservice orchestration could work well when executed externally. 

A combination of both strategies works too: you could have a dedicated workflow server to orchestrate cross-service processes, and individual micro services where certain logic is implemented through workflows rather than through code. Perhaps because a requirement was to have business rules implemented visually and to allow users to easily change existing or add new rules in the form of workflows.

The way I look at workflows is that they are just a programming language. and as with any programming language, you can use them in multiple apps in multiple ways.

Anyway, in order to be able to use Elsa in any app, you need to register all required services. The easiest way is to do `services.AddElsa()`, which will register all required services for basic operations (including `ISignaler`), as you did.

Regarding the `ReflectionTypeLoadException`, a commonly made mistake is to have multiple Elsa packages referenced of different versions.

Take a look at you .csproj files and make sure that any & all packages that start with Elsa are using the same package version.

If you're still getting the error even after you made sure all Elsa packages are of the same version, then feel free to zip your solution and attach it here and I will take a look.
